web-api/src/storage/postgres.rs

```diff
@@ -284,89 +297,154 @@ impl storage::Interest for Storage {
         Ok(UserInterests { positive, negative })
     }
 
-    async fn update_positive<F>(
-        &self,
-        doc_id: &DocumentId,
+    async fn upsert_cois(
+        tx: &mut Transaction<'_, Postgres>,
+        user_id: &UserId,
+        now: DateTime<Utc>,
+        cois: &[PositiveCoi],
+    ) -> Result<(), Error> {
+        let persist = match cois.len() {
+            0 => return Ok(()),
+            1 => true,
+            _ => false,
```

Comment:
Before non of the queries where prepared.

So I added a optimization to prepare queries if only one document was interacted with.

And then factored out the various functions.

As a side note this is about caching prepared queries, all queries are always prepared.

Tbh. just adding some `persist(true)` cases is kinda out of the scope of this PR, and there is no clear answer for if and where we should set a prepare limit so I set it to 1 as a  conservative default. Especially given that how our API is designed most times it will most likely be called once per interaction and as such with only 1 coi update.

If we only have well behaved clients probably not setting a limit would be best. But given that this APIs are publicly accessible it might make sense to limit it from preventing people from causing a performance degradation by "flushing out" the normally widely used statements. 
Maybe Postgres might be clever enough to handle this nicely, but I'm pretty sure any fire wall we do put in front will not help preventing this.

Reply:
:pushpin:  I generally want to have a generic push many values chunked by BIND_LIMIT abstraction.

Also performance wise it also makes sense to chunk on a much smaller basis then BIND_LIMIT.

Reply:
Times to insert 1 row in `center_of_interests`:
```
Planning Time: 0.014 ms
Execution Time: 0.187 ms
```
Times to insert 100 rows in `center_of_interests`:
```
Planning Time: 0.080 ms
Execution Time: 1.196 ms
```
With the following query:
```sql
EXPLAIN ANALYZE INSERT INTO "center_of_interest" ("coi_id", "user_id", "is_positive", "embedding", "view_count", "view_time_ms", "last_view") VALUES
('2b7bc6d6-b650-4436-8c48-6388e2e44291'::UUID, '0 user_id_coi_id_user_id_is_positive_embedding_view_count_view_time_ms', '1', '<embedding 128>', '4', '5', '2022-12-22 19:10:25-07'),
```
For the query with 100 rows the `coi_id`, `user_id`, and `embedding` are unique for each row.

With embeddings that have 384 values:
1 row
```
Planning Time: 0.017 ms
Execution Time: 0.235 ms
```
100 rows
```
Planning Time: 0.173 ms
Execution Time: 3.059 ms
```

Reply:
that is not testing/bench-marking what I'm referring to (see other comments)